pkg/datapath/iptables/iptables.go

```diff
@@ -1291,31 +1227,69 @@ func (m *IptablesManager) installHostTrafficMarkRule(prog iptablesInterface) err
 		"-m", "mark", "!", "--mark", matchFromIPSecEncrypt, // Don't match ipsec traffic
 		"-m", "mark", "!", "--mark", matchFromProxy, // Don't match proxy traffic
 		"-m", "comment", "--comment", "cilium: host->any mark as from host",
-		"-j", "MARK", "--set-xmark", markAsFromHost}, false)
+		"-j", "MARK", "--set-xmark", markAsFromHost})
 }
 
 // InstallRules installs iptables rules for Cilium in specific use-cases
 // (most specifically, interaction with kube-proxy).
-func (m *IptablesManager) InstallRules(ifName string, firstInitialization, install bool) (err error) {
-	quiet := firstInitialization
+func (m *IptablesManager) InstallRules(ctx context.Context, ifName string, firstInitialization, install bool) error {
+	backoff := backoff.Exponential{
+		Min:  20 * time.Second,
+		Max:  3 * time.Minute,
+		Name: "iptables-rules-installer",
+	}
+
+	maxAttempts := 3
+	attempt := 0
+
+	for {
+		attempt += 1
+		err := m.doInstallRules(ifName, firstInitialization, install)
+		if err == nil {
+			log.Info("Iptables rules installed")
+			return nil
+		}
+
+		if attempt == maxAttempts {
+			return fmt.Errorf("failed to install iptables rules: %w", err)
+		}
+
+		log.WithError(err).Warning("Failed to install iptables rules")
+		backoff.Wait(ctx)
+	}
+}
+
+func (m *IptablesManager) doInstallRules(ifName string, firstInitialization, install bool) error {
+	m.Lock()
+	defer m.Unlock()
 
 	// Make sure we have no old "backups"
-	m.removeRules(oldCiliumPrefix, true)
+	if err := m.removeRules(oldCiliumPrefix); err != nil {
+		return err
+	}
 
-	m.renameChains(oldCiliumPrefix, quiet)
+	if err := m.renameChains(oldCiliumPrefix); err != nil {
+		return err
+	}
 
 	// install rules if needed
 	if install {
-		err = m.installRules(ifName)
+		if err := m.installRules(ifName); err != nil {
+			return err
+		}
+
 		// copy old proxy rules over
 		match := ""
 		if firstInitialization {
 			match = "cilium-dns-egress"
 		}
-		m.copyProxyRules(oldCiliumPrefix+ciliumPreMangleChain, match)
+
+		if err := m.copyProxyRules(oldCiliumPrefix+ciliumPreMangleChain, match); err != nil {
+			return fmt.Errorf("cannot install proxy rules, disruption to traffic selected by L7 policy possible: %w", err)
+		}
 	}
 
-	if err != nil {
+	if err := m.removeRules(oldCiliumPrefix); err != nil {
```

Comment:
This already confused me in the v1.10 backport ... why does `doInstallRules()` call `m.removeRules(oldCiliumPrefix)` twice?